dali/python/nvidia/dali/plugin/mxnet.py

```diff
@@ -229,40 +241,60 @@ def __init__(self,
-            auto_reset,
-            last_batch_policy)
+        _DALIMXNetIteratorBase.__init__(self,
+                                        pipelines,
+                                        size,
+                                        reader_name,
+                                        fill_last_batch,
+                                        last_batch_padded,
+                                        auto_reset,
+                                        last_batch_policy,
+                                        prepare_first_batch=prepare_first_batch)
         self._squeeze_labels = squeeze_labels
         self._dynamic_shape = dynamic_shape
         # Use double-buffering of data batches
         self._data_batches = [[None] for i in range(self._num_gpus)]
         self._current_data_batch = 0
 
         self._first_batch = None
-        try:
-            self._first_batch = DALIGenericIterator.__next__(self)
-        except StopIteration:
-            assert False, "It seems that there is no data in the pipeline. This may happen if `last_batch_policy` is set to PARTIAL and the requested batch size is greater than the shard size."
-        # Set data descriptors for MXNet
-        self.provide_data = []
-        self.provide_label = []
-
-        category_names = {key : [] for key in self._output_categories}
-        for name, category in output_map:
-            category_names[category].append(name)
-        for i, data in enumerate(self._first_batch[0].data):
-            data_shape  = (data.shape[0] * self._num_gpus,) + data.shape[1:]
-            self.provide_data.append(mx.io.DataDesc(category_names[DALIGenericIterator.DATA_TAG][i], \
-                data_shape, data.dtype, layout=data_layout))
-        for i, label in enumerate(self._first_batch[0].label):
-            label_shape = (label.shape[0] * self._num_gpus,) + label.shape[1:]
-            self.provide_label.append(mx.io.DataDesc(category_names[DALIGenericIterator.LABEL_TAG][i], \
-                label_shape, label.dtype))
+        self._descriptors_populated = False
+        self._data_layout = data_layout
+        if self._prepare_first_batch:
+            try:
+                self._first_batch = DALIGenericIterator.__next__(self)
+            except StopIteration:
+                assert False, "It seems that there is no data in the pipeline. This may happen if `last_batch_policy` is set to PARTIAL and the requested batch size is greater than the shard size."
+
+    def __getattr__(self, key):
+        if key == 'provide_data' or key == 'provide_label':
+            # obtain the first batch to populate the metadata
+            try:
+                self._first_batch = DALIGenericIterator.__next__(self)
+            except StopIteration:
+                assert False, "It seems that there is no data in the pipeline. This may happen if `last_batch_policy` is set to PARTIAL and the requested batch size is greater than the shard size."
+        raise AttributeError
```

Comment:
Maybe it would be good to add a comment here, that those attributes is how we communicate with the MXNet and how you should implement an iterator for this framework.
Also, I think it should have a return value in non error code path to be compliant with the API.

Reply:
Done